source/RoslynAnalyzers/Testing/Integration/IntegrationTestForwardCancellationTokenToInvocationsAnalyzer.cs

```diff
@@ -0,0 +1,271 @@
+            {
+                // Avoid comparing to itself, or when there is no ct parameter, or when the last parameter is not a ct
+                if (SymbolEqualityComparer.Default.Equals(originalMethod, methodToCompare) ||
+                    methodToCompare.Parameters.Count(p => SymbolEqualityComparer.Default.Equals(p.Type, cancellationTokenType)) != 1 ||
+                    !SymbolEqualityComparer.Default.Equals(methodToCompare.Parameters[methodToCompare.Parameters.Length - 1].Type, cancellationTokenType))
+                {
+                    return false;
+                }
+
+                IMethodSymbol originalMethodWithAllParameters = (originalMethod.ReducedFrom ?? originalMethod).OriginalDefinition;
+                IMethodSymbol methodToCompareWithAllParameters = (methodToCompare.ReducedFrom ?? methodToCompare).OriginalDefinition;
+
+                // Ensure parameters only differ by one - the ct
+                if (originalMethodWithAllParameters.Parameters.Length != methodToCompareWithAllParameters.Parameters.Length - 1)
+                {
+                    return false;
+                }
+
+                // Now compare the types of all parameters before the ct
+                // The largest i is the number of parameters in the method that has fewer parameters
+                for (int i = 0; i < originalMethodWithAllParameters.Parameters.Length; i++)
+                {
+                    IParameterSymbol originalParameter = originalMethodWithAllParameters.Parameters[i];
+                    IParameterSymbol comparedParameter = methodToCompareWithAllParameters.Parameters[i];
+                    if (!SymbolEqualityComparer.Default.Equals(originalParameter.Type, comparedParameter.Type))
+                    {
+                        return false;
+                    }
+                }
+
+                // Overload is valid if its return type is implicitly convertable
+                var toCompareReturnType = methodToCompareWithAllParameters.ReturnType;
+                var originalReturnType = originalMethodWithAllParameters.ReturnType;
+                if (!toCompareReturnType.IsAssignableTo(originalReturnType))
+                {
+                    // Generic Task-like types are special since awaiting them essentially erases the task-like type.
+                    // If both types are Task-like we will warn if their generic arguments are convertable to each other.
+                    if (IsTaskLikeType(originalReturnType) && IsTaskLikeType(toCompareReturnType) &&
+                        originalReturnType is INamedTypeSymbol originalNamedType &&
+                        toCompareReturnType is INamedTypeSymbol toCompareNamedType &&
+                        TypeArgumentsAreConvertable(originalNamedType, toCompareNamedType))
+                    {
+                        return true;
+                    }
+
+                    return false;
+                }
+
+                return true;
```

Comment:
I suspect this could be simplified for someone reading the code. In general, prefer 'positive' vs 'negative' if statements. Theoy being, `if (aThing)` is easier to reason about and has less cognitive overhead than `if (!aThing)`.

In this case, I suspect you could do something like the following to reduce nesting?

```csharp
if (toCompareReturnType.IsAssignableTo(...)) {
  return true;
}

if (IsTaskLikeType(...) && IsTaskLikeType(...)) {
  ...
}
```

Reply:
Simplified as per suggested. The `if (IsTaskLikeType(...))` is updated to `return ...;`